lib: Implementing space saving algorithm in Gatekeeper

gogapp opened the pull request:
This PR implements space saving algorithm in Gatekeeper.

cjdoucette commented on Makefile:
```diff
@@ -30,7 +30,7 @@ include $(RTE_SDK)/mk/rte.vars.mk
 APP = gatekeeper
 
-SRCS-y := main/main.c
+SRCS-y := main/main.c 
```
Remove the extra space at the end of this line.

gogapp commented:
Hi mentors,

I have added the changes that you have requested. I have also merged the commits into a single meaningful commit. 

Please note that the implementation is still not fully complete and I am working on the TODOs present in the code.

cjdoucette commented on lib/space_saving.c:
```diff
@@ -0,0 +1,795 @@
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
```
All of the `case` statements should be at the same level of indentation as the `switch` statement.

cjdoucette commented on lib/space_saving.c:
```diff
@@ -0,0 +1,795 @@
+{
+	int ret;
+
+	switch(flow->proto) {
```
Add a space between `switch` and the parenthesis.

cjdoucette commented on lib/space_saving.c:
```diff
@@ -0,0 +1,795 @@
+	int ret;
+
+	switch(flow->proto) {
+		case ETHER_TYPE_IPv4:
```
Same thing here: add a space and bring everything inside of this `switch` back one level of indentation.

cjdoucette commented on lib/space_saving.c:
```diff
@@ -0,0 +1,795 @@
+	ct_hash_params->hash_func_init_val = 0;
+	ct_hash_params->reserved = 0;
+
+	char ct_name[64];
```
This should be declared at the top of the function, not in the middle.

cjdoucette commented on lib/space_saving.c:
```diff
@@ -0,0 +1,795 @@
+	 * after testing.
+	 */
+
+	if(ipv4_if_configured(iface)) {
```
Add a space after the `if` and before the parenthesis. Look through the rest of the code for similar cases.

gogapp commented:
I have added the code files for the final review and fixed the styling errors. For testing of the algorithms, I have created a different repository here. [](https://github.com/gogapp/GSOC-2018-tests)

cjdoucette commented:
Can you rebase your code to include one commit to add the space saving algorithm and one commit to add the RHHH algorithm? In other words, we shouldn't have a separate commit for fixing the style issues.

Also, the Gatekeeper code does not call the functions you are adding. To merge these changes we will need to have it actually integrated with Gatekeeper.

gogapp commented:
I have rebased the commits as requested.

To get the changes integrated with Gatekeeper, I need to think out ways to do this. Currently I am focusing on my College Placements. I will integrate the functions with Gatekeeper as soon as I am free.

AltraMayor commented:
Suggestion for where to integrate: `gt_proc()` at `if (pkt_info.priority <= 1) {`. That is, you'd feed all packets that are in accepted flows to the algorithms you implemented.
